src/Types/Interval.purs

```diff
@@ -380,161 +430,434 @@ after h (Interval { to: to' }) = upperBound h > to'
   let
-    lbound =
-      map
-        (if startIncl then slotToBeginPOSIXTime sc else slotToEndPOSIXTime sc)
-        start
-    ubound =
-      map
-        (if endIncl then slotToEndPOSIXTime sc else slotToBeginPOSIXTime sc)
-        end
+    slotLength = getSlotLength eraSummary
   in
-    Interval
-      { from: LowerBound lbound startIncl
-      , to: UpperBound ubound endIncl
-      }
-
--- | Convert a `Slot` to a `POSIXTimeRange` given a `SlotConfig`. Each `Slot`
--- | can be represented by an interval of time.
-slotToPOSIXTimeRange :: SlotConfig -> Slot -> POSIXTimeRange
-slotToPOSIXTimeRange sc slot =
-  interval (slotToBeginPOSIXTime sc slot) (slotToEndPOSIXTime sc slot)
-
--- | Get the starting `POSIXTime` of a `Slot` given a `SlotConfig`.
-slotToBeginPOSIXTime :: SlotConfig -> Slot -> POSIXTime
-slotToBeginPOSIXTime (SlotConfig { slotLength, slotZeroTime }) (Slot n) =
+    wrap $ relSlot * slotLength
+
+-- As justified in https://github.com/input-output-hk/ouroboros-network/blob/bd9e5653647c3489567e02789b0ec5b75c726db2/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Qry.hs#L461-L481
+-- Treat the upperbound as inclusive.
+-- | Returns the absolute time relative to some system start, not UNIX epoch.
+absTimeFromRelTime
+  :: EraSummary -> RelTime -> Either SlotToPosixTimeError AbsTime
+absTimeFromRelTime (EraSummary { start, end }) (RelTime relTime) = do
   let
-    msAfterBegin = uIntToBigInt n * slotLength
+    startTime = unwrap (unwrap start).time * factor
+    absTime = startTime + relTime -- relative to System Start, not UNIX Epoch.
+    -- If `EraSummary` doesn't have an end, the condition is automatically
+    -- satisfied. We use `<=` as justified by the source code.
+    -- Potential FIXME: note the hack that we don't have `end` for the current
+    -- era, if we did not do this, there could be issues going far into the
+    -- future?
+    endTime = maybe (absTime + one)
+      ((*) factor <<< unwrap <<< _.time <<< unwrap)
+      end
+  unless (absTime <= endTime) (throwError $ EndTimeLessThanTime $ wrap absTime)
+  pure $ wrap absTime
+
+--------------------------------------------------------------------------------
+-- POSIXTime (milliseconds) to
+-- Slot (absolute from System Start - see QueryM.SystemStart.getSystemStart)
+--------------------------------------------------------------------------------
+data PosixTimeToSlotError
+  = CannotFindTimeInEraSummaries AbsTime
+  | PosixTimeBeforeSystemStart POSIXTime
+  | StartTimeGreaterThanTime AbsTime
+  | EndSlotLessThanSlotOrModNonZero AbsSlot ModTime
+  | CannotConvertAbsSlotToSlot AbsSlot
+  | CannotGetBigIntFromNumber'
+
+derive instance Generic PosixTimeToSlotError _
+derive instance Eq PosixTimeToSlotError
+
+instance Show PosixTimeToSlotError where
+  show = genericShow
+
+-- | Converts a `POSIXTime` to `Slot` given an `EraSummariesQR` and
+-- | `SystemStartQR` queried from Ogmios.
+posixTimeToSlot
+  :: EraSummariesQR
```

Comment:
Maybe the `QR` suffix should be dropped from these type names since we use them in domain functions?